### PR TITLE
Feature/ Challenge based on rounds

### LINKED ISSUE #2399: process challenges based on round
- validate that create-challenge transaction is included in the block with exact round
- process challenges based on round created and round response submitted, make them expire based on current round as well

### PR DESCRIPTION
**Pull request by Jayashsatolia403**
## Tests
Tasks to complete before merging PR:
- [ ]  Ensure system tests are passing. If not [Run them manually](https://github.com/0chain/0chain/actions/workflows/system_tests.yml) to check for any regressions :clipboard:
- [ ]  Do any new system tests need added to test this change? do any existing system tests need updated? If so create a PR at [0chain/system_test](https://github.com/0chain/system_test)
- [ ]  Merge your system tests PR to master AFTER merging this PR

## Associated PRs (Link as appropriate):
- blobber:
- gosdk:
- system_test:
- zboxcli:
- zwalletcli:
- Other: ...

### CODE REVIEW

File: code/go/0chain.net/smartcontract/storagesc/challengepool.go [hunk 1]
```diff
@@ -192,16 +193,16 @@ func (ssc *StorageSmartContract) getChallengePool(allocationID datastore.Key,
...
 	_, err = ssc.getChallengePool(allocationID, balances)
-	switch err {
-	case util.ErrValueNotPresent:
+	switch {
+	case errors.Is(err, util.ErrValueNotPresent):
 		cp = newChallengePool()
 		cp.TokenPool.ID = challengePoolKey(ssc.ID, allocationID)
 		return cp, nil
-	case nil:
+	case err == nil:
```
Comment (dabasov) [inline code/go/0chain.net/smartcontract/storagesc/challengepool.go]: nil is good

File: code/go/0chain.net/smartcontract/storagesc/challengepool.go [hunk 2]
```diff
@@ -192,16 +193,16 @@ func (ssc *StorageSmartContract) getChallengePool(allocationID datastore.Key,
...
 // challenge pool for a client don't saving it
 func (ssc *StorageSmartContract) newChallengePool(allocationID string,
-	creationDate, expiresAt common.Timestamp, balances cstate.StateContextI) (
+	balances cstate.StateContextI) (
 	cp *challengePool, err error) {
 
 	_, err = ssc.getChallengePool(allocationID, balances)
-	switch err {
-	case util.ErrValueNotPresent:
+	switch {
```
Comment (dabasov) [inline code/go/0chain.net/smartcontract/storagesc/challengepool.go]: empty switch is not good, you better break it in ifs

**Review by dabasov — CHANGES_REQUESTED** [review]

**Review by dabasov — APPROVED** [review]
lgtm

File: code/go/0chain.net/smartcontract/stakepool/edb_stakepool.go [hunk 3]
```diff
@@ -21,6 +22,9 @@ func (sp *StakePool) EmitStakePoolBalanceUpdate(
 		dp := sp.Pools[id]
 		dpu := dbs.NewDelegatePoolUpdate(id, pId, pType)
 		dpu.Updates["balance"] = dp.Balance
+
+		logging.Logger.Info("EmitStakePoolBalanceUpdate", zap.Any("dpu", dpu))
```
Comment (dabasov) [inline code/go/0chain.net/smartcontract/stakepool/edb_stakepool.go]: remove log please or change level to debug

File: code/go/0chain.net/smartcontract/storagesc/allocation.go [hunk 4]
```diff
@@ -1136,6 +1141,17 @@ func (sc *StorageSmartContract) settleOpenChallengesAndGetPassRates(
 	passRates = make([]float64, 0, len(alloc.BlobberAllocs))
 
 	allocChallenges, err := sc.getAllocationChallenges(alloc.ID, balances)
+	if err != nil {
+		if err == util.ErrValueNotPresent {
+			allocChallenges = &AllocationChallenges{}
+			allocChallenges.AllocationID = alloc.ID
+		} else {
```
Comment (dabasov) [inline code/go/0chain.net/smartcontract/storagesc/allocation.go]: if ErrorNotFound return it unchanged

File: code/go/0chain.net/smartcontract/storagesc/allocation.go [hunk 5]
```diff
@@ -1362,6 +1395,11 @@ func (sc *StorageSmartContract) finalizeAllocation(
 			"saving allocation: "+err.Error())
 	}
 
+	logging.Logger.Info("Jayash Updating allocation to DB",
```
Comment (dabasov) [inline code/go/0chain.net/smartcontract/storagesc/allocation.go]: remove logging

**Review by dabasov — COMMENTED** [review]
c

**Comment by service-0chain** [comment]
Manual system tests [failure] with the following config
| config  | value |
| ------------- | ------------- |
| tests skipped  | false  |
| system-tests  |   |
| gosdk  | NOT OVERRIDDEN  |
| zboxcli  |   |
| zwalletcli  |   |
| 0chain  | sprint-1.11  |
| blobber  |   |
| authorizer  |   |
| 0box  |   |
| 0dns  |   |

**Pull request merged**